resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesSuite.scala

```diff
@@ -366,4 +371,49 @@ private[spark] class KubernetesSuite extends SparkFunSuite with BeforeAndAfter {
         s" with correct contents."), "Job did not find the file as expected.")
     }
   }
+
+  test("Use external URI provider") {
+    val externalUriProviderWatch = new ExternalUriProviderWatch(minikubeKubernetesClient)
+    Utils.tryWithResource(minikubeKubernetesClient.services()
+        .withLabel("spark-app-name", "spark-pi")
+        .watch(externalUriProviderWatch)) { _ =>
+      val args = Array(
+        "--master", s"k8s://https://${Minikube.getMinikubeIp}:8443",
+        "--deploy-mode", "cluster",
+        "--kubernetes-namespace", NAMESPACE,
+        "--name", "spark-pi",
+        "--executor-memory", "512m",
+        "--executor-cores", "1",
+        "--num-executors", "1",
+        "--jars", HELPER_JAR_FILE.getAbsolutePath,
+        "--class", SPARK_PI_MAIN_CLASS,
+        "--conf", "spark.ui.enabled=true",
+        "--conf", "spark.testing=false",
+        "--conf", s"spark.kubernetes.submit.caCertFile=${clientConfig.getCaCertFile}",
+        "--conf", s"spark.kubernetes.submit.clientKeyFile=${clientConfig.getClientKeyFile}",
+        "--conf", s"spark.kubernetes.submit.clientCertFile=${clientConfig.getClientCertFile}",
+        "--conf", "spark.kubernetes.executor.docker.image=spark-executor:latest",
+        "--conf", "spark.kubernetes.driver.docker.image=spark-driver:latest",
+        "--conf", "spark.kubernetes.submit.waitAppCompletion=false",
+        "--conf", s"${DRIVER_SERVICE_MANAGER_TYPE.key}=${ExternalSuppliedUrisDriverServiceManager.TYPE}",
+        EXAMPLES_JAR_FILE.getAbsolutePath)
+      SparkSubmit.main(args)
+      val sparkMetricsService = getSparkMetricsService("spark-pi")
+      expectationsForStaticAllocation(sparkMetricsService)
+      assert(externalUriProviderWatch.annotationSet.get)
+      val driverService = minikubeKubernetesClient
+        .services()
+        .withLabel("spark-app-name", "spark-pi")
+        .list()
+        .getItems
+        .asScala(0)
+      assert(driverService.getMetadata.getAnnotations.containsKey(ANNOTATION_PROVIDE_EXTERNAL_URI),
+          "External URI request annotation was not set on the driver service.")
+      // Unfortunately we can't check the correctness of the actual value of the URI, as it depends
+      // on the driver submission port set on the driver service but we remove that port from the
+      // service once the submission is complete.
```

Comment:
I think this service cleanup after the job is submitted should be part of the plugin point, rather than the core always opinionatedly deleting the service after the job has submitted.

Alternatively the contract could be that any created k8s resources need to have owner references to the service so that when Spark deletes the service the dependent resources are cleaned up too